RecoHI/HiJetAlgos/plugins/HiFJRhoFlowModulationProducer.cc

```diff
@@ -0,0 +1,251 @@
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescriptionFiller.h"
+#include "DataFormats/HeavyIonEvent/interface/EvtPlane.h"
+#include "DataFormats/JetReco/interface/JetCollection.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidateFwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+
+#include "TF1.h"
+#include "TH1.h"
+#include "TMath.h"
+
+#include <algorithm>
+#include <cmath>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+class HiFJRhoFlowModulationProducer : public edm::stream::EDProducer<> {
+public:
+  explicit HiFJRhoFlowModulationProducer(const edm::ParameterSet&);
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void beginStream(edm::StreamID) override;
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  void endStream() override;
+
+  bool doEvtPlane_;
+  bool doFreePlaneFit_;
+  bool doJettyExclusion_;
+  int evtPlaneLevel_;
+  edm::EDGetTokenT<reco::JetView> jetTag_;
+  edm::EDGetTokenT<reco::PFCandidateCollection> pfCandsToken_;
+  edm::EDGetTokenT<reco::EvtPlaneCollection> evtPlaneToken_;
+
+  float* hiEvtPlane;
+};
+
+HiFJRhoFlowModulationProducer::HiFJRhoFlowModulationProducer(const edm::ParameterSet& iConfig)
+    : doEvtPlane_(iConfig.getParameter<bool>("doEvtPlane")),
+      doFreePlaneFit_(iConfig.getParameter<bool>("doFreePlaneFit")),
+      doJettyExclusion_(iConfig.getParameter<bool>("doJettyExclusion")),
+      evtPlaneLevel_(iConfig.getParameter<int>("evtPlaneLevel")),
+      jetTag_(consumes<reco::JetView>(iConfig.getParameter<edm::InputTag>("jetTag"))),
+      pfCandsToken_(consumes<reco::PFCandidateCollection>(iConfig.getParameter<edm::InputTag>("pfCandSource"))),
+      evtPlaneToken_(consumes<reco::EvtPlaneCollection>(iConfig.getParameter<edm::InputTag>("EvtPlane"))) {
+  produces<std::vector<double>>("rhoFlowFitParams");
+}
+
+// ------------ method called to produce the data  ------------
+void HiFJRhoFlowModulationProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<reco::PFCandidateCollection> pfCands;
+  iEvent.getByToken(pfCandsToken_, pfCands);
+
+  edm::Handle<reco::EvtPlaneCollection> evtPlanes;
+  if (doEvtPlane_) {
+    iEvent.getByToken(evtPlaneToken_, evtPlanes);
+    if (evtPlanes.isValid()) {
+      for (unsigned int i = 0; i < evtPlanes->size(); ++i) {
+        hiEvtPlane[i] = (*evtPlanes)[i].angle(evtPlaneLevel_);
+      }
+    }
+  }
+
+  edm::Handle<reco::JetView> jets;
+  if (doJettyExclusion_)
+    iEvent.getByToken(jetTag_, jets);
+
+  int nFill = 0;
+
+  double eventPlane2 = -100;
+  double eventPlane3 = -100;
+
+  constexpr int nParamVals = 9;
+  auto rhoFlowFitParamsOut = std::make_unique<std::vector<double>>(nParamVals, 1e-6);
```

Comment:
Sorry can't figure that out, `auto rhoFlowFitParamsOut = std::make_unique<std::array<double, nParamVals>>` not seem to work

Reply:
what was the error?

Reply:
Probably just doing something wrong
` error: pointer to a function used in arithmetic [-Werror=pointer-arith]
   rhoFlowFitParamsOut[0] = 0;
   error: assignment of read-only location '**rhoFlowFitParamsOut'
   rhoFlowFitParamsOut[0] = 0;
`

Reply:
but this looks more like about ```(*rhoFlowFitParamsOut)[0]``` vs ```rhoFlowFitParamsOut[0]```

Reply:
Same error if I change to `(*rhoFlowFitParamsOut)[0]`

Reply:
I may be missing what's being tried on your side.
The following compiles for me:
```c++
  auto arrUP = std::make_unique<std::array<double, 9>>();
  (*arrUP)[2] = 3;
  std::cout<<(*arrUP)[2]<<std::endl;
```

Reply:
I was missing the `()` at the end when the array is defined, thank you

Reply:
Also I was wondering what is the right way to `iEvent.put` this array if it expects a vector

Reply:
I may have not considered the fact that this is to be saved in the event.
@makortel @Dr15Jones do we have support for std::arrays saved directly in edm::Event (even if transient)?

Reply:
`ticl::Trackster` contains `std::array` s and it has a ROOT dictionary.

Reply:
indeed, array as a part of a class works (also used in PFCandidate).
However, there are no StdDictionaries .
@stepobr please update the produces call to write an array.
You may need to add ```std::array<double,9>``` (or what's the size) in  DataFormats/StdDictionaries/src/classes_def_others.xml
and also add an include for ```<array>```  in DataFormats/StdDictionaries/src/classes_others.h

Reply:
I think ROOT has built in direct support for std::array which is why we do not have any explicit dictionary definitions for them even though they are used as member data.

Reply:
even better, no need to fiddle with DataFormats/StdDictionaries/src/

Reply:
So should I directly `iEvent.put(rhoFlowFitParamsOut);` ?

Reply:
move is expected to be better; IIUC, the issue was that the matching ```produces``` call should have a std::array instead of a vector

Reply:
@slava77 I'd like to get back to this. So when I change the matching `produces` to `std::array', as well as the `consumes` in `CSJetProducer` I get this:
`
An exception of category 'DictionaryNotFound' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=HiFJRhoFlowModulationProducer label='hiFJRhoFlowModulationProducer'
   [2] Calling ProductRegistryHelper::addToRegistry, checking dictionaries for produced types
Exception Message:
No data dictionary found for the following classes:

  edm::Wrapper<std::array<double,9> >
  std::array<double,9>

`

I tried to add the array in `StdDictionaries` but that didn't help.

Reply:
> I tried to add the array in `StdDictionaries` but that didn't help.

this seems to disagree with 
https://github.com/cms-sw/cmssw/pull/30057#discussion_r445126814
@Dr15Jones it looks like the support to stream array types directly is not that simple (if even present)

Reply:
Adding the array to `StdDictionaries` gives an error during compilation:
`Error: It is not necessary to explicitly select class array<double,9>. I/O is supported for it transparently.`

Reply:
The symptoms (ROOT saying it has a dictionary, framework that it has not) smell similar to https://github.com/cms-sw/cmssw/pull/27861. I'll take a look.

Reply:
The cause is different from #27861, and may take some time to resolve. I opened an issue for it https://github.com/cms-sw/cmssw/issues/30493.

Reply:
In case the resolution will take longer than ~today, I'm willing to accept the current vector solution for this product.
With an assumption/note that it can/should be updated once the framework has support for array.

Reply:
Almost certainly will take longer than today.